### PR TITLE
drivers: sensor: Add ROHM BH1750 driver

### PR DESCRIPTION
**Pull request by gportay**
Dear Maintainers,

This is my first driver, and I wonder if you are interested in having it upstream.

The sensor has several working modes, and I have picked up the mode "one-time high resolution mode"; there are also five other modes that I can implement (including low resolution, continuous...). Do you want I implement all of them?

I have also a sample code, tell me if it needs to be part of the PR.

The code is based on the Bosch BME280 source code.

Kind Regards,
Gaël

--- 

This adds support for ROHM BH1750 the i2c light sensor.

Signed-off-by: Gaël PORTAY <gael.portay@gmail.com>

### CODE REVIEW

File: drivers/sensor/bh1750/bh1750.c [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+	const struct bh1750_config *cfg = dev->config;
+
+	return device_is_ready(cfg->i2c.bus) ? 0 : -ENODEV;
+}
+
+static inline int bh1750_read(const struct device *dev, uint8_t *buf, int size)
+{
+	const struct bh1750_config *cfg = dev->config;
+
+	return i2c_read(cfg->i2c.bus, buf, size, cfg->i2c.addr);
```
Comment (MaureenHelm) [inline drivers/sensor/bh1750/bh1750.c:46]: Please use `i2c_read_dt` instead

File: drivers/sensor/bh1750/bh1750.c [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+static inline int bh1750_write(const struct device *dev, uint8_t val)
+{
+	const struct bh1750_config *cfg = dev->config;
+	uint8_t buf = val;
+
+	return i2c_write(cfg->i2c.bus, &buf, sizeof(buf), cfg->i2c.addr);
```
Comment (MaureenHelm) [inline drivers/sensor/bh1750/bh1750.c:54]: Please use `i2c_write_dt` instead

File: drivers/sensor/bh1750/bh1750.c [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+	}
+
+	ret = bh1750_write(dev, BH1750_RESET);
+	if (ret < 0) {
+		LOG_DBG("Reset failed: %d", ret);
+		return ret;
+	}
+
+	/* Wait for the sensor to be ready */
+	k_sleep(K_MSEC(1));
```
Comment (MaureenHelm) [inline drivers/sensor/bh1750/bh1750.c:152]: How were all of the wait times chosen?

File: drivers/sensor/bh1750/bh1750.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+#define BH1750_POWER_ON                            0x01
+#define BH1750_RESET                               0x07
+#define BH1750_CONTINUOUSLY_HIGH_RESOLUTION_MODE   0x10
+#define BH1750_CONTINUOUSLY_HIGH_RESOLUTION_MODE_2 0x11
+#define BH1750_CONTINUOUSLY_LOW_RESOLUTION_MODE    0x13
+#define BH1750_ONE_TIME_HIGH_RESOLUTION_MODE       0x20
+#define BH1750_ONE_TIME_HIGH_RESOLUTION_MODE_2     0x21
+#define BH1750_ONE_TIME_LOW_RESOLUTION_MODE        0x23
+#define BH1750_CHANGE_MEASUREMENT_TIME_HIGH        0x40
+#define BH1750_CHANGE_MEASUREMENT_TIME_LOW         0x60
```
Comment (MaureenHelm) [inline drivers/sensor/bh1750/bh1750.h:27]: Please move to the .c file and remove the .h file

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
> This is my first driver, and I wonder if you are interested in having it upstream.
> 

Yes, thank you for the contribution!

> The sensor has several working modes, and I have picked up the mode "one-time high resolution mode"; there are also five other modes that I can implement (including low resolution, continuous...). Do you want I implement all of them?

It's not required. If you want to add them, please do that in a follow up PR after this PR gets merged.

> 
> I have also a sample code, tell me if it needs to be part of the PR.

We are trying to reduce the number of driver-specific samples. If you're interested in helping create a generic light sensor sample similar to `samples/sensor/accel_polling/` and `samples/sensor/magn_polling/` and enabling the sample for this sensor, it would be a welcome addition.
> 
> The code is based on the Bosch BME280 source code.

If you've copied code from somewhere, please keep the original copyright and add your own copyright in a new line.

Please add this sensor tests/drivers/build_all/sensor to ensure the driver gets built in CI

File: drivers/sensor/bh1750/bh1750.c (hunk 3 above)
Comment (gportay) [reply drivers/sensor/bh1750/bh1750.c:169]: I do not recall for that one specifically.

Other is taken from datasheet

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Looks good overall, thank you.

Please review and address the failed twister checks.

File: dts/bindings/sensor/rohm,bh1750-i2c.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: ROHM integrated light sensor
+
+compatible: "rohm,bh1750"
+
+include: i2c-device.yaml
```
Comment (MaureenHelm) [inline dts/bindings/sensor/rohm,bh1750-i2c.yaml:7]:
```suggestion
include: [sensor-device.yaml, i2c-device.yaml]
```

See #50929

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
> Please review and address the failed twister checks.

This still needs to be addressed

File: drivers/sensor/bh1750/bh1750.c [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+		 * Low Byte is "10010000"
+		 *
+		 * (2^15 + 2^9 + 2^8 + 2^7 + 2^4) / 1.2 = 28067 [lx]
+		 *
+		 * Which means:
+		 *
+		 * lx = Slx / 1.2
+		 */
+		tmp = data->raw_val / 1.2;
+		ret = sensor_value_from_double(val, tmp);
```
Comment (gportay) [inline drivers/sensor/bh1750/bh1750.c:134]: @MaureenHelm: as for [51269](https://github.com/zephyrproject-rtos/zephyr/pull/51269#discussion_r997168142)

> You don't need floating point math for this

do you prefer the former solution with long long unsigned?

``` c
static int bh1750_channel_get(const struct device *dev,
			      enum sensor_channel chan,
			      struct sensor_value *val)
{
	struct bh1750_data *data = dev->data;
	long long unsigned tmp;

	switch (chan) {
	case SENSOR_CHAN_LIGHT:
		tmp = 1000000UL * data->raw_val / 1.2;
		val->val1 = tmp / 1000000;
		val->val2 = tmp % 1000000;
		break;
	default:
		return -EINVAL;
	}

	return 0;
}
```

Comment (gportay) [reply drivers/sensor/bh1750/bh1750.c:134]: @MaureenHelm do you want me to switch back to `long long unsigned` for that driver as well?

**Comment by gportay** [comment]
@MaureenHelm ping?

**Comment by gportay** [comment]
I am closing the MR, since another driver for this hardware was merged in #43086.

**Pull request closed without merging**